### PR TITLE
Implement translation system in Odyssey

### PR DESCRIPTION
**Pull request by francoislehoux-okta**
### Description
- [x] - Replace `optionalLabel` with un-customizable test (`Optional`).
- [x] - Rename ThemeProvider to OdysseyProvider
- [x] - Initial `OdysseyTranslationProvider` context provider
- [x]   - Possible to pass `language` (language code) to `OdysseyTranslationProvider` or else defaults to `en`
- [x]   - `useTranslation` hook to render the correct value using a key
- [x] Wire properties -> ts pipeline
- [x] Load languages via ts
- [x] Wire globalization-core properties pipeline (track request [here](https://oktainc.atlassian.net/browse/OKTA-609109))
- [x] Make useTranslation type safe (https://react.i18next.com/latest/typescript)
- [x] Remove all english leaks
- [x] default to browser language
- [x] Documentation
- [x] Need to determine key name convention with odyssey team
- [x] Add .ts translations to .gitignore

### Screenshots
`Default/EN`:
Screen Shot 2023-05-09 at 3 33 46 PM

`FR`:
Screen Shot 2023-05-09 at 3 34 03 PM

Type-safe:
Screen Shot 2023-05-10 at 12 30 49 PM

### CODE REVIEW

File: packages/odyssey-react-mui/package.json [hunk 1]
```diff
@@ -66,11 +73,12 @@
...
+    "build:i18n": "node scripts/properties-to-json.js bundle",
     "build:clean": "rm -rf dist",
     "build:icons": "svgr ../odyssey-icons/src --out-dir ./src/iconDictionary",
     "build:source": "NODE_ENV=production BROWSERSLIST_ENV=modern babel src --out-dir dist --extensions .js,.jsx,.ts,.tsx --source-maps",
     "build:types": "tsc --project tsconfig.production.json",
-    "build": "npm-run-all --sequential build:clean --parallel build:types build:source",
+    "build": "npm-run-all --sequential build:clean build:i18n --parallel build:types build:source",
     "dev:source": "yarn build:source --watch",
     "dev:types": "yarn build:types --watch",
     "dev": "npm-run-all --parallel dev:source dev:types",
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/package.json:84]: Should we call `build:i18n` for `dev` as well? @KevinGhadyani-Okta

File: packages/odyssey-react-mui/package.json [hunk 2]
```diff
@@ -66,11 +73,12 @@
     "react-dom": ">=17 <19"
   },
   "scripts": {
+    "build:i18n": "node scripts/properties-to-json.js bundle",
     "build:clean": "rm -rf dist",
     "build:icons": "svgr ../odyssey-icons/src --out-dir ./src/iconDictionary",
     "build:source": "NODE_ENV=production BROWSERSLIST_ENV=modern babel src --out-dir dist --extensions .js,.jsx,.ts,.tsx --source-maps",
     "build:types": "tsc --project tsconfig.production.json",
-    "build": "npm-run-all --sequential build:clean --parallel build:types build:source",
+    "build": "npm-run-all --sequential build:clean build:i18n --parallel build:types build:source",
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/package.json:81]: I believe `build:i18n` needs to be ran before `build:source` as the `ts` versions need to be available first

File: packages/odyssey-react-mui/scripts/properties-to-json.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+exports.handler = async (argv) => {
+  const { resourcePath, targetJsonPath } = argv;
+
+  await convertPropertiesToJson({
+    resourcePath,
+    targetJsonPath,
+  });
+};
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/scripts/properties-to-json.js:70]: This was taken from globalization-core

File: packages/odyssey-react-mui/scripts/properties-to-json.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      const targetFile = join(
+        propertiesTargetDir,
+        filename.replace(extension, ".ts")
+      );
+      ensureFileSync(targetFile);
+      writeFileSync(
+        targetFile,
+        "/* eslint-disable header/header */\nexport const translation = " +
+          JSON.stringify(obj) +
+          ";"
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/scripts/properties-to-json.js:45]: Disabling the header/header rule for the generated files. Also had to add the path to these to `.prettierignore`

File: packages/odyssey-react-mui/src/OdysseyI18n.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import i18n from "i18next";
+import { initReactI18next } from "react-i18next";
+
+import { translation as en } from "./properties/ts/odyssey-react-mui";
+import { translation as fr } from "./properties/ts/odyssey-react-mui_fr";
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/src/OdysseyI18n.ts:17]: Are supported languages will have to be imported as such once translated

File: packages/odyssey-react-mui/scripts/properties-to-json.js (hunk 4 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/scripts/properties-to-json.js:45]: Just add the header. It's copy-paste, but that's the easiest way to ensure it's there. We even do it in our generated icons.

Unless these are generated at compile-time and not comitted, then that's different.

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties [hunk 6]
```diff
@@ -0,0 +1,7 @@
+field.error.screen.reader.text = Erreur
+field.label.optional.text = Optionnel
+severity.success = succès
+severity.warning = avertissement
+severity.info = info
+severity.error = erreur
+toast.close.text = fermé
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:7]: I will be deleting this file toward the end as that will be added by the translation partners

File: packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties [hunk 7]
```diff
@@ -0,0 +1,7 @@
+field.error.screen.reader.text = Error
+field.label.optional.text = Optional
+severity.success = success
+severity.warning = warning
+severity.info = info
+severity.error = error
+toast.close.text = close
```
Comment (francoislehoux-okta) [inline packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties:7]: Need input about naming of key/value pairs

File: packages/odyssey-react-mui/scripts/properties-to-json.js (hunk 4 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/scripts/properties-to-json.js:45]: They will be, trying to add these to git ignore

File: packages/odyssey-react-mui/scripts/properties-to-json.js (hunk 3 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/scripts/properties-to-json.js:70]: Let's put a comment at the top of the file and link to the file in `globalization-core`. That way future devs will know where to look.

File: packages/odyssey-react-mui/src/@types/i18next.d.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+import "i18next";
+import { translation as en } from "../properties/ts/odyssey-react-mui";
+
+declare module "i18next" {
+  interface CustomTypeOptions {
+    resources: {
+      translations: typeof en;
+    };
+  }
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/@types/i18next.d.ts:21]: :thumbsup: Slick!

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties (hunk 6 above)
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:7]: This is just temporary for your own testing right? Once we merge this in, there'll be some config file from Globalization that tells them how to add properties files?

File: packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties (hunk 7 above)
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties:7]: We don't put this as `en` in the `translations/` folder?

File: packages/odyssey-react-mui/src/properties/ts/odyssey-react-mui.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+/* eslint-disable header/header */
+export const translation = {"field.error.screen.reader.text":"Error","field.label.optional.text":"Optional","severity.success":"success","severity.warning":"warning","severity.info":"info","severity.error":"error","toast.close.text":"close"};
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/properties/ts/odyssey-react-mui.ts:2]: Are we committing converted `.properties` translations to the repo?

I don't mind it, just curious. When they add `.properties` files, I'd like to generate these TS files in that step. Is that how you've set it up?

File: packages/odyssey-react-mui/src/OdysseyI18n.ts (hunk 5 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/OdysseyI18n.ts:17]: Ah, so this is a manual process? Is there a way to look at the `translations/` dir and automate that process as part of the TS file creation? I'm sure we could write in a simple Node.js script to do it.

File: packages/odyssey-react-mui/src/OdysseyProvider.tsx [hunk 10]
```diff
@@ -14,13 +14,16 @@ import { memo, ReactElement } from "react";
...
 import { OdysseyCacheProvider } from "./OdysseyCacheProvider";
 import { OdysseyThemeProvider } from "./OdysseyThemeProvider";
+import { OdysseyTranslationProvider } from "./OdysseyTranslationProvider";
 
-const ThemeProvider = ({ children }: { children: ReactElement }) => (
+const OdysseyProvider = ({ children }: { children: ReactElement }) => (
   <OdysseyCacheProvider>
-    <OdysseyThemeProvider>{children}</OdysseyThemeProvider>
+    <OdysseyThemeProvider>
+      <OdysseyTranslationProvider>{children}</OdysseyTranslationProvider>
```
[the remark this answers is not in the document]
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/OdysseyProvider.tsx:22]: For sure. We can do this in a future PR.

File: packages/odyssey-react-mui/src/OdysseyTranslationProvider.types.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  "pt-BR", // Portuguese (Brazil)
+  "ro", // Romanian
+  "ru", // Russian
+  "sv", // Swedish
+  "th", // Thai
+  "tr", // Turkish
+  "uk", // Ukrainian
+  "zh-CN", // Chinese (PRC)
+  "zh-TW", // Chinese
+] as const;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/OdysseyTranslationProvider.types.ts:40]: I'm wondering if we can also generate this list based on what's in `translations/` + `en`. Or is this used by Globalization?

File: packages/odyssey-react-mui/src/OdysseyTranslationProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+}: {
+  children: ReactElement;
+  language?: SupportedLanguages;
+}) {
+  useEffect(() => {
+    // Defaults to the browser's language if available otherwise `en` will be used
+    i18n.changeLanguage(language || window.navigator.language);
+  }, [language]);
+
+  return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/OdysseyTranslationProvider.tsx:32]: This is great! Thanks for implementing it in a way where it doesn't require a page reload to change!

**Comment by KevinGhadyani-Okta** [comment]
Great work on this PR! I'd like to get this in without translation overrides and move those to a separate PR.

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties [hunk 13]
```diff
@@ -0,0 +1,7 @@
+field.error.screen.reader.text = Erreur
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:1]: Should this be `screenreader.text`?

```suggestion
field.error.screenreader.text = Erreur
```

File: packages/odyssey-react-mui/scripts/properties-to-json.js (hunk 3 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/scripts/properties-to-json.js:70]: This works? @KevinGhadyani-Okta 
```
// Part of this has been copied over from @okta/ui-build-tools' own internal node scrip
// https://github.com/okta/ui-build-tools/blob/master/packages/clis/i18n/properties-to-json.js
```

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties (hunk 6 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:7]: Screen Shot 2023-05-10 at 2 35 03 PM

@KevinGhadyani-Okta

File: packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties (hunk 7 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/src/properties/odyssey-react-mui.properties:7]: Negative as that is the "master" that you guys will own (english). Everything under `translations/` is owned by the translation provider and globalization core. Correct @nikhilvenkatraman-okta ?

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties (hunk 13 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:1]: Great idea

File: packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties (hunk 6 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/src/properties/translations/odyssey-react-mui_fr.properties:7]: For testing purposes, if I DO have time, I will add the ok_PL (pseudoloc) script/language to this.

**Comment by KevinGhadyani-Okta** [comment]
> Should we call `build:i18n` for `dev` as well? @KevinGhadyani-Okta  

This is more like `generate:i18n` right? Because it's generating TS files, not just building stuff, but it's generating source files (same as icons).

**Review by KevinGhadyani-Okta — APPROVED** [review]
Approving, but I'd like to rename the command to `generate:i18n` instead of `build:i18n`. I'm fine committing the TS files and running this script as-needed since we do the same thing with icons. One less thing to build when developing.

**Comment by francoislehoux-okta** [comment]
> 

Will adjust in the follow-up PR

File: packages/odyssey-react-mui/README.md [hunk 14]
```diff
@@ -72,16 +74,34 @@ const YourAppRoot = ({ children }) => (
...
+
+### Overrides
+
+It is possible to provide `nonce`, `themeOverride` and `languageCode` to the underlying components via props:
+
+```jsx
+import { OdysseyProvider } from "@okta/odyssey-react-mui";
+import { theme } from "./themeOverride";
+
+const nonce = K6ybeaqlwaAKk0hQF?hE2qZ0Ar3M=G4Bpr-XFk1x4XbL3dH3T5=mk0aXeWX9ifn9;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/README.md:98]:
```suggestion
const nonce = "K6ybeaqlwaAKk0hQF?hE2qZ0Ar3M=G4Bpr-XFk1x4XbL3dH3T5=mk0aXeWX9ifn9";
```

File: packages/odyssey-react-mui/package.json [hunk 15]
```diff
@@ -47,17 +49,22 @@
     "@types/jest-axe": "^3.5.1",
     "@types/react": "^17.0.30",
     "@types/react-dom": "^17.0.5",
+    "@types/yargs": "^17",
     "babel-plugin-import": "^1.13.5",
     "eslint": "^7.27.0",
+    "fs-extra": "10.1.0",
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/package.json:55]: Pretty sure we shouldn't be using `fs-extra`.

Is this something we can remove?

File: packages/odyssey-react-mui/scripts/properties-to-ts.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      }
+
+      const targetFile = join(
+        propertiesTargetDir,
+        filename.replace(extension, ".ts")
+      );
+      ensureFileSync(targetFile);
+      writeFileSync(
+        targetFile,
+        "export const translation = " + JSON.stringify(obj) + ";"
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/scripts/properties-to-ts.js:45]:
```suggestion
        `export const translation = ${JSON.stringify(obj)};`
```

File: packages/odyssey-react-mui/package.json (hunk 15 above)
Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/package.json:55]: Investigating as part of the follow up: https://github.com/okta/odyssey/pull/1791

**Pull request merged**